src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java

```diff
@@ -0,0 +1,235 @@
+		this.file = Objects.requireNonNull(file);
+
+		if (!Files.exists(file))
+			throw new NoSuchFileException(file.toString());
+		if (Files.size(file) < 4096 * 2)
+			throw new IllegalArgumentException("File size must be at least 4kiB, is this file corrupt?");
+		FileChannel raf = FileChannel.open(file, StandardOpenOption.READ);
+
+		locations = ByteBuffer.allocate(4096);
+		raf.read(locations);
+		locations.flip();
+		locations2 = locations.asIntBuffer();
+
+		timestamps = ByteBuffer.allocate(4096);
+		raf.read(timestamps);
+		timestamps.flip();
+		timestamps2 = timestamps.asIntBuffer();
+
+		chunks = new Chunk[32 * 32];
+		chunksUpToDate = new boolean[32 * 32];
+		for (int i = 0; i < 1024; i++) {
+			int chunkPos = locations2.get(i) >>> 8;
+			int chunkLength = locations2.get(i) & 0xFF;
+			if (chunkPos > 0) {
+				/* i & 31 retrieves the last 5 bit which store the x coordinate */
+				chunks[i] = new Chunk(i & 31, i >> 5, timestamps2.get(i), raf, chunkPos, chunkLength);
+			} else
+				chunks[i] = null;
+			chunksUpToDate[i] = true;
+		}
+	}
+
+	/**
+	 * Returns the {@link Chunk} at the given coordinate
+	 *
+	 * @param x
+	 *            the x coordinate relative to this region file's origin
+	 * @param z
+	 *            the z coordinate relative to this region file's origin
+	 * @return the chunk at that coordinate or {@code null} if the chunk does not exist
+	 * @author piegames
+	 */
+	public Chunk getChunk(int x, int z) {
+		return chunks[(x & 31) | (z << 5)];
+	}
+
+	public Chunk getChunk(int i) {
+		return chunks[i];
+	}
+
+	/**
+	 * Set a chunk at the given coordinate
+	 * 
+	 * @param x
+	 *            the x coordinate relative to this region file's origin
+	 * @param z
+	 *            the z coordinate relative to this region file's origin
+	 * @param chunk
+	 *            The chunk to set. Use {@code null} to remove it.
+	 * @author piegames
+	 */
+	public synchronized void setChunk(int x, int z, Chunk chunk) {
+		setChunk((x & 31) | (z << 5), chunk);
+	}
+
+	public synchronized void setChunk(int i, Chunk chunk) {
+		chunks[i] = chunk;
+		chunksUpToDate[i] = false;
+	}
+
+	/**
+	 * Write all chunks that got changed to disk, update the file's header (chunk locations and timestamps) and truncate the file at the end
+	 * 
+	 * @author piegames
+	 */
+	public synchronized void writeChanges() throws IOException {
+		try (FileChannel raf = FileChannel.open(file, StandardOpenOption.WRITE, StandardOpenOption.CREATE);) {
+			BitSet usedSectors = new BitSet();
+			usedSectors.set(0, 2); /* Set the first two sectors as used since they always ares */
+			List<Integer> toSave = new ArrayList<>();
+			for (int i = 0; i < 32 * 32; i++) {
+				if (!chunksUpToDate[i]) {
+					toSave.add(i);
+					continue;
+				}
+				if (chunks[i] == null)
+					continue;
+				int chunkPos = locations2.get(i) >>> 8;
+				int chunkLength = locations2.get(i) & 0xFF;
+				usedSectors.set(chunkPos, chunkPos + chunkLength);
```

Comment:
usedSectors is being populated with old state. This will not cause corruption (I think), but it does mean that any mutated chunks will be likely pushed to the end of the file even though they could fit in there old location. (wasted space)

Reply:
`usedSectors` is populated with the old state for chunks that haven't changed (see the continue in line 137). If the modified chunks haven't grown, they'll be reinserted there afterwards.